Source/Core/DolphinLibretro/Input.cpp

```diff
@@ -495,36 +521,38 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
-    gcCStick->SetControlExpression(1, "`" + devAnalog + ":Y1+`");         // Down
-    gcCStick->SetControlExpression(2, "`" + devAnalog + ":X1-`");         // Left
-    gcCStick->SetControlExpression(3, "`" + devAnalog + ":X1+`");         // Right
-    gcDPad->SetControlExpression(0, "Up");                                // Up
-    gcDPad->SetControlExpression(1, "Down");                              // Down
-    gcDPad->SetControlExpression(2, "Left");                              // Left
-    gcDPad->SetControlExpression(3, "Right");                             // Right
-    gcTriggers->SetControlExpression(0, "L");                             // L
-    gcTriggers->SetControlExpression(1, "R");                             // R
-    gcTriggers->SetControlExpression(2, "`" + devAnalog + "Trigger0+`");  // L-trigger Analog
-    gcTriggers->SetControlExpression(3, "`" + devAnalog + "Trigger1+`");  // R-trigger Analog
+    gcButtons->SetControlExpression(0, "A");                          // A
+    gcButtons->SetControlExpression(1, "B");                          // B
+    gcButtons->SetControlExpression(2, "X");                          // X
+    gcButtons->SetControlExpression(3, "Y");                          // Y
+    gcButtons->SetControlExpression(4, "R");                          // Z
+    gcButtons->SetControlExpression(5, "Start");                      // Start
+    gcMainStick->SetControlExpression(0, "`" + devAnalog + ":Y0-`");  // Up
+    gcMainStick->SetControlExpression(1, "`" + devAnalog + ":Y0+`");  // Down
+    gcMainStick->SetControlExpression(2, "`" + devAnalog + ":X0-`");  // Left
+    gcMainStick->SetControlExpression(3, "`" + devAnalog + ":X0+`");  // Right
+    gcCStick->SetControlExpression(0, "`" + devAnalog + ":Y1-`");     // Up
+    gcCStick->SetControlExpression(1, "`" + devAnalog + ":Y1+`");     // Down
+    gcCStick->SetControlExpression(2, "`" + devAnalog + ":X1-`");     // Left
+    gcCStick->SetControlExpression(3, "`" + devAnalog + ":X1+`");     // Right
+    gcDPad->SetControlExpression(0, "Up");                            // Up
+    gcDPad->SetControlExpression(1, "Down");                          // Down
+    gcDPad->SetControlExpression(2, "Left");                          // Left
+    gcDPad->SetControlExpression(3, "Right");                         // Right
+    gcTriggers->SetControlExpression(0, "`" + devAnalog + ":Trigger0+`|(L2&!`" + devAnalog +
+                                            ":Trigger0+`)");  // L-trigger Full Press
+    gcTriggers->SetControlExpression(1, "`" + devAnalog + ":Trigger1+`|(R2&!`" + devAnalog +
+                                            ":Trigger1+`)");  // R-trigger Full Press
+    gcTriggers->SetControlExpression(2,
+                                     "`" + devAnalog + ":Trigger0+`|L3");  // L-trigger Soft Press
+    gcTriggers->SetControlExpression(3,
+                                     "`" + devAnalog + ":Trigger1+`|R3");  // R-trigger Soft Press
```

Comment:
mixed triggers should have a digital control expression for 0-1 and an analog one for 2-3. I don't see the point of using such a complex expression. also why `L3`/`R3` ? triggers are either `L1` or `L2`

Reply:
ok I assume this part here was the problem : https://github.com/libretro/dolphin/blob/master/Source/Core/InputCommon/ControllerEmu/ControlGroup/MixedTriggers.cpp#L54-L60
were the analog inputs being ignored when using digital input for 0-1 ?

Reply:
would changing the code in `MixedTriggers.cpp` like this work with the simpler control expressions? :
```    // Apply threshold:
    if (button_value > threshold)
    {
      // Activate button:
      *digital |= bitmasks[i];
    }
    if (analog_value > threshold)
    {
      // Fully activate analog:
      analog_value = 1.0;
    }
```
I don't have a suitable gamepad to test.

Reply:
regarding the L2/L3 part, I see you were tricking the game into thinking that it was a half press, by only setting the analog input to full while keeping the digital input cleared? kinda weird but I guess that is ok if it does work for the games you tested. L2/L3 aren't used for anything else afterall.